manual/manual/cmds/intf-c.etex

```diff
@@ -477,21 +475,34 @@ floating-point numbers.}
 
 \subsection{ss:c-outside-head}{Pointers outside the heap}
 
-Any word-aligned pointer to an address outside the heap can be safely
-cast to and from the type "value". This includes pointers returned by
-"malloc", and pointers to C variables (of size at least one word)
-obtained with the \verb'&' operator.
-   %%% FIXME will change in 4.02.0 (?)
-
-Caution: if a pointer returned by "malloc" is cast to the type "value"
-and returned to OCaml, explicit deallocation of the pointer using
-"free" is potentially dangerous, because the pointer may still be
-accessible from the OCaml world.  Worse, the memory space deallocated
-by "free" can later be reallocated as part of the OCaml heap; the
-pointer, formerly pointing outside the OCaml heap, now points inside
-the OCaml heap, and this can crash the garbage collector.  To avoid
-these problems, it is preferable to wrap the pointer in a OCaml block
-with tag "Abstract_tag" or "Custom_tag".
+In earlier versions of OCaml, it was possible to use
+word-aligned pointers to addresses outside the heap as OCaml values,
+just by casting the pointer to type "value".  Starting with OCaml
+4.11, this usage is obsolete and will stop being supported soon.
```

Comment:
We do still allow some kinds of out-of-heap pointer, though, and I think it's worth documenting their usage (since for some applications, the overhead of having an associated heap block will be too high).  What about:
- replace "will stop" with "will in general stop";
- after the code examples, add a new paragraph: "For advanced uses, word-aligned pointers to addresses outside the heap may still be cast to the type `value`, but every such pointer must point to an address where the word immediately prior is as created by the `Caml_out_of_heap_header` macro.  This macro currently produces a block header marked black, but this may change in future releases."

Reply:
See the discussion in #9564.  What you wrote is not enough for the naked pointer to be safe for ad-hoc polymorphic primitives (equality, etc): the data that it points to must agree with the tag and size of the header.  Using `Caml_out_of_heap_header` requires "insider" knowledge about object representation that go beyond what this chapter describes (namely: the FFI).  So I don't want to document `Caml_out_of_heap_header` in this chapter.